[WIP] New narratives tutorial (#139)
nextstrain/docs.nextstrain.org: src/tutorials/narratives-how-to-write.rst

===================
Writing a narrative
===================

The following will walk you through writing a Nextstrain narrative.
We'll start with a basic 2-slide narrative and then gradually add further slides with the intention
of describing the functionality of narratives. The aim is for you to feel comfortable with
the technical aspects of creating and troubleshooting narratives.

For an introduction to the concept of Narratives, please see `Communicating Results Using Narratives </guides/communicate/narratives-intro.html>`__ or take a look at this example narrative,  `Twenty years of West Nile virus <https://next.nextstrain.org/narratives/twenty-years-of-WNV>`__.

> This tutorial will rely on the new `Narratives Debugger <https://dev.nextstrain.org/edit/narratives>`__ which allows you to drag & drop the markdown files we'll introduce here and see an instant summary of the narrative.

Reviewer: I think we should consistently capitalize Markdown as the proper noun it is.

```suggestion
This tutorial will rely on the new `Narratives Debugger <https://dev.nextstrain.org/edit/narratives>`__ which allows you to drag & drop the Markdown files we'll introduce here and see an instant summary of the narrative.
```

There are several examples of "markdown" in other places.

Author: Fixed - https://github.com/nextstrain/docs.nextstrain.org/compare/03a1276b5dfebbb8d81a3447d64cd31239b9f4e4..477e27d73e3cde37e40c662e0b3aea0a9c4cb987

.. TODO- update links above once functionality in nextstrain.org

.. contents:: Sections in this document
  :local:
  :depth: 2

.. note::

  This guide uses the narratives debugger on nextstrain.org; while you can also use this by running Auspice on your computer there are some differences - see :ref:`below for more details <auspice_vs_nextstrain>`.

Prerequisites
=============

In order to make sense of this tutorial we are going to need:

* A (basic) understanding of writing markdown `(here is good introduction) <https://learnxinyminutes.com/docs/markdown/>`__.

> * Familiarity with the general concept of Nextstrain Narratives (see `Communicating Results Using Narratives </guides/communicate/narratives-intro.html>`__ to get started).

Reviewer: This sort of link should be `:doc:` references, e.g.

```suggestion
* Familiarity with the general concept of Nextstrain Narratives (see :doc:`/guides/communicate/narratives-intro` to get started).
```

You can omit the link text if you want to use the document's title.

There's several more examples of this mistake throughout the document.

Note also that while links like `/guides/communicate/narratives-intro.html` will appear to work in the review site, they won't on the live site.

Author: > Note also that while links like /guides/communicate/narratives-intro.html will appear to work in the review site, they won't on the live site.

Ouch! That's quite the papercut. Changing them now...

Author: Fixed up - thanks!

..
  TODO- update links above once debugger functionality in nextstrain.org

.. image :: ../images/narratives_debugger_screenshot.jpg
   :alt: Screenshot of the example narrative loaded in the narratives debugger and a preview of the opening slide

We can see the titles of the three slides & their associated datasets (see the following section for details on how titles are defined), and hovering over these should show you the full appearance of the slide.
To the right of the titles are their associate datasets, each with a series of icons next to them representing the main + sidecar JSONs associated with this dataset.
The icons represent whether the dataset exists on nextstrain.org -- in this case, they are all green (success) or grey (not attempted).
When debugging a narrative it's easy to make syntax errors resulting in datasets which don't load -- you could try changing the URLs in the markdown file and dragging the file back onto the debugger to see this.
Clicking on the title of a slide will jump into the narrative at that slide, or you can click the "View Narrative" button to start from the beginning.

.. note::

>   The narrative file you drag onto the debugger remains within your browser and is not sent to the nextstrain server.

Reviewer: ```suggestion
  The narrative file you drag onto the debugger remains within your browser and is not sent to the Nextstrain server.
```

Author: Fixed - https://github.com/nextstrain/docs.nextstrain.org/compare/03a1276b5dfebbb8d81a3447d64cd31239b9f4e4..477e27d73e3cde37e40c662e0b3aea0a9c4cb987

Hopefully the above section gives you enough to get started writing your own narratives.
There are more technical details to follow, but if you want to get started we encourage trying it out!
There are plenty of ways to approach the task, but we find the following workflow tends to work well:

.. graphviz::
    :align: center

    strict digraph {
        rankdir="LR";
        auspice [shape="tab" style="filled" fillcolor="#c7e9b4" label="nextstrain.org/...\nto choose desired \nview of data"]
        md [shape="note" style="filled" fillcolor="#41b6c4" label="Narrative file\nwe are writing\n(markdown)"]
        debugger [shape="tab" style="filled" fontcolor="white" fillcolor="#225ea8" label="Narratives debugger\nto test narrative\nas we go"]
        
        auspice -> md [label="copy\nURL" fontcolor="#7fcdbb" fillcolor="#7fcdbb" color="#7fcdbb"]
        md -> auspice [label="repeat" fontcolor="#7fcdbb" fillcolor="#7fcdbb" color="#7fcdbb" splines=curved]
        md -> debugger [label="drag &\ndrop" fontcolor="#1d91c0" fillcolor="#1d91c0" color="#1d91c0"]
        debugger -> md [label="repeat" fontcolor="#1d91c0" fillcolor="#1d91c0" color="#1d91c0" splines=curved]

>     }

Reviewer: I think would be nice to use a sans-serif font here, like the surrounding body text and our other graphviz diagrams.

Author: Done - https://github.com/nextstrain/docs.nextstrain.org/compare/03a1276b5dfebbb8d81a3447d64cd31239b9f4e4..477e27d73e3cde37e40c662e0b3aea0a9c4cb987

#. Any translators are then listed, encoded in the same format as the authors but using keys ``translators`` and ``translatorLinks``.
#. The abstract, defined by ``abstract`` is a string which will be rendered as markdown [#f1]_.
#. When the narrative was first created (``date``) and when it was most recently updated (``updated``) is then displayed.
#. Finally, any applicable license is shown, as defined by ``license`` and ``licenseLink``.
#. The ``dataset`` is required and defines the data view in the right hand side of screen (see :ref:`see below <linking-view-to-url>`).

Normal slides
-------------

The rest of the markdown file defines one or more slides, where each slide is defined by a level 1 heading which is also a link to a dataset on nextstrain.org [#f2]_ and a section of markdown which represents the slide's content:

.. graphviz::
    :align: center

    strict digraph {
        title [shape="box" style="filled" fillcolor="#feb24c" label="# [Slide Title](nextstrain.org dataset URL)"]
        body [shape="box" style="filled" fillcolor="#fd8d3c" label="Markdown content of slide"]
        title -> body [style="invis"]

>     }

Reviewer: It's kinda weird to me to use a graphviz diagram to represent the Markdown contents of a single slide. My first impression upon reading was that each node (`title` and `body`) were representing different _slides_ instead of parts of a single slide. Why not use a `.. code-block:: markdown` block like later?

Author: Ha! Yeah I got carried away with my new graphviz expertise

Author: Done - https://github.com/nextstrain/docs.nextstrain.org/compare/03a1276b5dfebbb8d81a3447d64cd31239b9f4e4..477e27d73e3cde37e40c662e0b3aea0a9c4cb987

The following example of a narrative slide may make this difference clearer:

.. code-block:: markdown

  # [Which dataset is this?](https://nextstrain.org/community/inrb-drc/ebola-nord-kivu)

  The dataset displayed here differs if you run it on localhost (auspice)
  or through nextstrain.org.

  When viewed on nextstrain.org this dataset is sourced from
  [this GitHub repository](https://github.com/inrb-drc/ebola-nord-kivu)
  as it is using our
  [community sharing functionality](https://docs.nextstrain.org/en/latest/guides/share/community-builds.html).

  To work when running on auspice locally, we would need to have a dataset
  with the following filename: `community_inrb-drc_ebola-nord-kivu.json`.

This complexity encouraged us to build the interactive debugger we have used in this tutorial and we encourage to use that approach rather than developing your narratives using Auspice locally & then trying to share them through nextstrain.org!

Where can I find some example narratives to use as inspiration?
---------------------------------------------------------------

> All of the narratives written by the core nextstrain team are available in `this GitHub repo <github.com/nextstrain/narratives>`__.

Reviewer: ```suggestion
All of the narratives written by the core Nextstrain team are available in `this GitHub repo <https://github.com/nextstrain/narratives>`__.
```

Author: Fixed - https://github.com/nextstrain/docs.nextstrain.org/compare/03a1276b5dfebbb8d81a3447d64cd31239b9f4e4..477e27d73e3cde37e40c662e0b3aea0a9c4cb987

Future directions
-----------------

The app used to test narratives is in its infancy.

> We hope to incrementally add features such as markdown editing, changing the dataset view settings, and the ability to publish the narrative to nextstrain groups.

Reviewer: ```suggestion
We hope to incrementally add features such as Markdown editing, changing the dataset view settings, and the ability to publish the narrative to :doc:`Nextstrain Groups </learn/groups/index>`.
```

Author: 👍 https://github.com/nextstrain/docs.nextstrain.org/compare/03a1276b5dfebbb8d81a3447d64cd31239b9f4e4..477e27d73e3cde37e40c662e0b3aea0a9c4cb987

The eventual aim is to be able to write & publish an entire narrative from within the app, without needing to know any specifics of the markdown language behind it.

Bugs? Improvements? Suggestions?
================================

The debugger (as of 2022-01-08) is in a beta-release phase. Please get in touch if you have suggestions or find bugs!
You can submit an `issue on GitHub <https://github.com/nextstrain/nextstrain.org/>`__ or make a post on our `discussion forum <discussion.nextstrain.org/>`__.

--------------

.. rubric:: Footnotes

.. [#f1] The provided string is actually rendered as ``### <abstract>``, so the first line will appear as a h3 heading!
  Multiple line strings are possible in YAML and we suggest using these.

.. [#f2] In practice, the URL path component is what is used to access the dataset relative to the current hostname, however for historical reasons the protocol and hostname must be either ``http://localhost:4000`` or ``https://nextstrain.org``. 

> .. [#f3] This isn't great from a file-size point of view, and the markdown file isn't nice to look at itself, but it may be the best current solution!

Reviewer: I think a nice user experience would be to have `nextstrain remote upload` (and any other narrative uploading/editing clients) auto-embed images this way. It makes the narrative self-contained, which is really nice. (Think back to the bad old days when most Word docs didn't embed images in them by default, but merely referenced them on the filesystem. It was bad.)

Author: I, happily, do not remember those dark days 😂 

Would be a nice user experience for locally referenced images for sure - the alternatives get pretty complicated pretty fast

Reviewer: Yeah, locally available images would be auto-embedded but hotlinked/remote images could be left alone. With a web-based editor, dragging and dropping an image would directly embed (or upload to somewhere we control and then hotlink, but I think this has lots of downsides).

  is dominated by samples from Europe and the Americas, whereas the
  basal (older) cases were typically from Africa.

  # [The rise and fall of Influenza clades](https://nextstrain.org/flu/seasonal/h3n2/ha/12y?d=tree,frequencies&p=full)

  We switch now to a completely different viral phylogeny - that of
  Influenza H3N2 over 12 years. This switch was made by defining a
  different dataset in the URL associated with this slide
  (https://nextstrain.org/flu/seasonal/h3n2/ha/12y?d=tree,frequencies&p=full).
  It is this URL which specifies that we should view both the tree & the
  frequencies panel 👉

  P.S. This phylogenetic structure is very different, more latter-like than
  comb like, a phenomena which is linked to antigenic evolution
  ([Bedford, Rambaut & Pascual, 2012](https://bmcbiol.biomedcentral.com/articles/10.1186/1741-7007-10-38)).

You can see a copy of this Markdown file `on GitHub <https://github.com/nextstrain/narratives/blob/master/how-to-write_basics.md>`__, and you can see the resulting narrative displayed in Nextstrain at `nextstrain.org/narratives/how-to-write/basics <https://nextstrain.org/narratives/how-to-write/basics>`__.

> This tutorial will make use of our narratives debugger, which is available at `nextstrain.org/debug-narrative <https://nextstrain.org/edit/narratives>`__.

Author: doh! Changed the actual URL but not the rendered URL 🤦 Fixed now

The following example of a narrative slide may make this difference clearer:

.. code-block:: markdown

  # [Which dataset is this?](https://nextstrain.org/community/inrb-drc/ebola-nord-kivu)

  The dataset displayed here differs if you run it on localhost (auspice)
  or through nextstrain.org.

  When viewed on nextstrain.org this dataset is sourced from
  [this GitHub repository](https://github.com/inrb-drc/ebola-nord-kivu)
  as it is using our
  [community sharing functionality](https://docs.nextstrain.org/en/latest/guides/share/community-builds.html).

  To work when running on auspice locally, we would need to have a dataset
  with the following filename: `community_inrb-drc_ebola-nord-kivu.json`.

This complexity encouraged us to build the interactive debugger we have used in this tutorial and we encourage to use that approach rather than developing your narratives using Auspice locally & then trying to share them through nextstrain.org!

Where can I find some example narratives to use as inspiration?
---------------------------------------------------------------

> All of the narratives written by the core Nextstrain team are available in `this GitHub repo <github.com/nextstrain/narratives>`__.

Reviewer: Ah, I noticed this again during re-review but somehow missed adding the comment for it :facepalm: 

```suggestion
All of the narratives written by the core Nextstrain team are available in `this GitHub repo <https://github.com/nextstrain/narratives>`__.
```

Author: Not a problem -- fixed up in https://github.com/nextstrain/docs.nextstrain.org/compare/9425839a96b2cffe9481c7ccf2a4158de08f4909..ea98849a3ee13ec31dfce5a1986449bff43316b7

Future directions
-----------------

The app used to test narratives is in its infancy.
We hope to incrementally add features such as Markdown editing, changing the dataset view settings, and the ability to publish the narrative to Nextstrain Groups.
The eventual aim is to be able to write & publish an entire narrative from within the app, without needing to know any specifics of the Markdown language behind it.

Bugs? Improvements? Suggestions?
================================

The debugger (as of November 2022) is in a beta-release phase. Please get in touch if you have suggestions or find bugs!

> You can submit an `issue on GitHub <https://github.com/nextstrain/nextstrain.org/>`__ or make a post on our `discussion forum <discussion.nextstrain.org/>`__.

Author: https://github.com/nextstrain/docs.nextstrain.org/compare/477e27d73e3cde37e40c662e0b3aea0a9c4cb987..9425839a96b2cffe9481c7ccf2a4158de08f4909

--------------

.. rubric:: Footnotes

.. [#f1] The provided string is actually rendered as ``### <abstract>``, so the first line will appear as a h3 heading!
  Multiple line strings are possible in YAML and we suggest using these.

.. [#f2] In practice, the URL path component is what is used to access the dataset relative to the current hostname, however for historical reasons the protocol and hostname must be either ``http://localhost:4000`` or ``https://nextstrain.org``. 

> 

Author: I think it may have been removed - I noticed this reviewing a separate PR which changes the narrative parsing. I think the current implementation is that the hostname is completely unused?

> 

Reviewer 2: #146 contains 5feeb49bf982f6807c4a386138980a594c220ed2 which removes the footnote.
